### PR TITLE
Add intra-population edges to test circuit, fix external edge merge bug

### PR DESCRIPTION
**Pull request by cattabiani**
## Summary

Adds intra-population edges (A→A, B→B) to the `split_subcircuit` test circuit and fixes a bug in `_copy_filtered_edges` exposed by the new edge topology.

## Test data changes

- **A→A edges**: A3→A4, A4→A3, A1→A2, A0→A4, A5→A1
- **B→B edges**: B0→B2, B3→B4
- **C→C**: intentionally omitted to test the case where no intra-population edge file exists

Updated `circuit_config.json`, `circuit_config_subcircuit.json`, and regenerated `edges.h5`. All test assertions updated to reflect the new external populations and edge counts.

### The problem

`_copy_filtered_edges` used to concatenate all source DataFrames into one:
```python
src_concat = pd.concat(id_mapping.data[write_edge_config.src_mapping].values())
```

This is conceptually wrong. The sources come from different ID spaces (e.g., the parent's `external_A` population vs the local `A` population). When two sources happen to share the same local node ID, the concatenated DataFrame has duplicate index values, and `.loc[id]` returns multiple rows — corrupting the edge file.

### Concrete example

When extracting c4 from c3_c2_c1, the `external_A` population has two sources:
- `"external_A"`: node 2 (inherited from parent) → new_id 0
- `"A"`: node 2 (newly removed) → new_id 1

The concat produces index `[2, 2]`. When processing the parent's `external_A__C` edge (src=2, tgt=3), the lookup `src_concat.loc[2]` returns both rows, appending `[0, 1]` to source_node_id but only `[3]` to target_node_id.

### The fix

Removed `src_concat` entirely. Each input now resolves its own source DataFrame directly via `source_filter`. For single-source populations (the common case), the single entry is used directly. For multi-source populations, an assertion enforces that `source_filter` is always provided.

### Why it wasn't triggered before

The bug requires intra-population edges (A→A) so that removed A nodes become `external_A` via the same-population path. Without them, external nodes were only discovered through inter-population edges, and the local ID collision never occurred.

c2_c1
c3_c1
c3_c2_c1
c4_c1
c4_c2_c1
c4_c3_c1
c4_c3_c2_c1
Original_c1

### CODE REVIEW

**Comment by mgeplf** [comment]
> _copy_filtered_edges used to concatenate all source DataFrames into one:
> src_concat = pd.concat(id_mapping.data[write_edge_config.src_mapping].values())

I don't see `src_concat` in the PR; so this is a confusing description.

I do see `src_stacked = pd.concat(id_mapping.data[write_edge_config.src_mapping].values())`; however, *this was just added in your last PR*:  https://github.com/openbraininstitute/brainbuilder/pull/53/changes#diff-bcab5e7d80484b653fadfe2504e8f0242003bc496f8983c082e1d3d2d0e6186fR287

Does that mean we are knowingly merging code with known bugs?  That doesn't sit well with me.

**Comment by cattabiani** [comment]
@mgeplf 

> Does that mean we are knowingly merging code with known bugs? That doesn't sit well with me.

I was in a catch 22 situation:

- pr A adds a bug
- pr B (on top of A) adds tests -> discovers bug

If I add the bug fix in pr A I do not get the tests (they are in pr B) to be sure that it is fixed. 
Did I have to combine A and B? Then you complain that it is a monolithic PR and I need to split it

Let me know what is the correct how to and I will follow that next time

**Comment by mgeplf** [comment]
> If I add the bug fix in pr A I do not get the tests (they are in pr B) to be sure that it is fixed.
> Did I have to combine A and B? Then you complain that it is a monolithic PR and I need to split it
>Let me know what is the correct how to and I will follow that next time

You can still add the tests in A, to show that it handles the new case.
WRT monolithic PRs, they need to be split in sensical ways; if a new feature is added, and there is some refactoring that goes with it, it can be in the same PR.  If there are 2 different bug fixes and a new feature, that's 3 PRs.
IMO, what isn't a good idea is doing refactoring, introducing a known bug, and then fixing it in another PR - that means the repo is in a known buggy state due to new code being introduced.  We're lucky that this doesn't have a lot of turn over, or multiple developers working on it.

File: brainbuilder/utils/sonata/split_population.py [hunk 1]
```diff
@@ -310,15 +307,19 @@ def _copy_filtered_edges(
             with h5py.File(input_path, "r") as h5in:
                 orig_edges = h5in["edges"][src_edge_name]
 
-                # Filter source node IDs by source key if specified
+                # Resolve source mapping for this input
```
Comment (mgeplf) [inline brainbuilder/utils/sonata/split_population.py]: superfluous

File: brainbuilder/utils/sonata/split_population.py [hunk 2]
```diff
@@ -310,15 +307,19 @@ def _copy_filtered_edges(
...
+                src_sources = id_mapping.data[write_edge_config.src_mapping]
+
                 if source_filter is not None:
-                    source_df = id_mapping.data[write_edge_config.src_mapping].get(source_filter)
-                    if source_df is not None:
-                        filtered_sgids = source_df.index.to_numpy()
-                    else:
-                        filtered_sgids = sgids_new
+                    source_df = src_sources[source_filter]
```
Comment (mgeplf) [inline brainbuilder/utils/sonata/split_population.py]: was trying to understand what `source_filter` was, and it seems that there isn't a description in `WriteEdgeConfig::inputs` of what the tuple values are; perhaps it should have a comment or maybe it should be a dataclass as well.

I find it strange to call it a "filter", when it seems to be a src_population name?

**Review by mgeplf — COMMENTED** [review]
couple small things, otherwise LGTM

File: brainbuilder/utils/sonata/split_population.py (hunk 1 above)
Comment (cattabiani) [reply brainbuilder/utils/sonata/split_population.py]: removed

File: brainbuilder/utils/sonata/split_population.py (hunk 2 above)
Comment (cattabiani) [reply brainbuilder/utils/sonata/split_population.py]: ok. improved the name

**Review by mgeplf — APPROVED** [review]
